src/typecheck/mod.rs

```diff
@@ -639,6 +639,23 @@ fn type_check_<S, E>(
             envs.insert(x.clone(), ty_let);
             type_check_(state, envs, lin, linearizer, strict, rt, ty)
         }
+        Term::LetBlock(defs, rest) => {
+            for (id, term) in defs.iter() {
+                let ty_let = binding_type(term.as_ref(), &mut env, global, state.table, strict);
+                type_check_(state, global, lin, linearizer.scope(ScopeId::Left), strict, term, Closure::new(ty_let.clone(), env.clone()))?;
+                env.insert(id.clone(), Closure::new(ty_let, env.clone()));
+            }
```

Comment:
Same thing as with evaluation, you need to typecheck this in a recursive environment. You can look at the `RecRecord` case to see how to it. We may even separate this typechecking of mutually recursive bindings in a separate helper as I expect the code to be almost the same for let blocks and recursive records.